### ISSUE TITLE
`PYI041`: Extend rule beyond parameter type-annotations

### ISSUE THREAD

**Issue by euclio** [astral-sh/ruff#24453]
### Summary

With the following code:

```python
from dataclasses import dataclass
from pydantic import BaseModel

class Foo(BaseModel):
    bar: int | float

@dataclass
class Baz:
    quux: int | float

def xyzzy(_quux: int | float):
    pass
```

The lint only fires on the function parameters. I would expect it to fire for the classes as well.

```
$ uvx ruff check --isolated --select PYI041 test.py
PYI041 [*] Use `float` instead of `int | float`
  --> test.py:11:18
   |
 9 |     quux: int | float
10 |
11 | def xyzzy(_quux: int | float):
   |                  ^^^^^^^^^^^
12 |     pass
   |
help: Remove redundant type
```

Playground link: https://play.ruff.rs/ab4dd23b-71e9-495e-b254-93af58068bec

### Version

ruff 0.15.9 (724ccc1ae 2026-04-02)

**Comment by MichaReiser** [comment]
This was an intentional decision when the rule was added in https://github.com/astral-sh/ruff/pull/5722 because of a mypy incompatibility https://github.com/PyCQA/flake8-pyi/issues/299 and to match upstream. 

I tried to reproduce the linked issue around mypy's type promotion in a recent mypy version, but failed. I also was unable to install mypy 0.990 locally to get a better sense of what the issue was.

Either way, I agree that the special case doesn't just apply to parameters but in all type annotation positions, as discussed in https://discuss.python.org/t/clarifying-the-float-int-complex-special-case/54018/31.

@AlexWaygood do you have any concerns around expanding the rule's scope (this should be a preview only change)

**Comment by AlexWaygood** [comment]
I'd prefer to leave this rule as it is. The [typing spec](https://typing.python.org/en/latest/spec/special-types.html#special-cases-for-float-and-complex) still only specifies what the behaviour should be for parameter annotations. To try to make the rule sound, ty applies this special case much more broadly, and other type checkers also apply the special case somewhat more broadly than what the spec says. But I wouldn't say that there is currently any consensus or unanimity in exactly how far the special case should be interpreted.

There's also a chance that ty [might change its behaviour](https://github.com/astral-sh/ty/issues/3241) on how it interprets the special case, since it seems to have been quite confusing for our users.

Even for type checkers that do have a broader interpretation of the `int`/`float` special case, moreover, I think there's an argument that the special case makes much more intuitive sense in the context of parameter annotations, and can be eveh more confusing in situations like attribute annotations.
